### PR TITLE
tests: added tests for download client

### PR DESCRIPTION
**Pull request by chmielewskiandreas**
Initial test environment for testing the download client. Currently just a
subset is implemented to test the download client using coap.

Signed-off-by: Andreas Chmielewski <andreas.chmielewski@grandcentrix.net>

### CODE REVIEW

**Comment by gopiotr** [comment]
@jtguggedal @hakonfam @sigvartmh @simensrostad please perform review or point more suitable person to do this ;)

File: tests/subsys/net/lib/download_client/src/mock/dl_coap.h [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+#ifndef _DL_COAP_H_
+#define _DL_COAP_H_
+
+#include <zephyr.h>
+
+struct override_ret_values_t {
```
Comment (rlubos) [inline tests/subsys/net/lib/download_client/src/mock/dl_coap.h:11]: struct names shouldn't contain `_t` suffix.

File: tests/subsys/net/lib/download_client/src/main.c [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+static void dl_coap_start(struct download_client *client)
+{
+	static const char host[] = "coap://10.1.0.10";
+	int err;
+
+	memset(client, 0, sizeof(struct download_client));
+
+	err = download_client_init(client, download_client_callback);
+	zassert_ok(err, NULL);
+	// FIXME: possible race condition; give time to client_init to start the dl thread
```
Comment (rlubos) [inline tests/subsys/net/lib/download_client/src/main.c:93]: Maybe it'd be enough to align `ZTEST_THREAD_PRIORITY` with download thread prio, and `k_yield()` here? This should give the download thread a chance to run.

File: tests/subsys/net/lib/download_client/src/main.c [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+	for (i = 0; i < len; i++) {
+		z_ztest_returns_value(func, val[i]);
+	}
+}
+
+static bool wait_for_event(enum download_client_evt_id event, uint8_t seconds)
+{
+	int i;
+
+	for (i = 0; i < (20 * seconds); i++) {
```
Comment (rlubos) [inline tests/subsys/net/lib/download_client/src/main.c:67]: Is the math correct here? 20 iterations with 10 ms wait time gives 200ms, not full second.

File: tests/subsys/net/lib/download_client/testcase.yaml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+tests:
+  net.lib.download_client:
+    tags: aws fota
+    platform_allow: nrf9160dk_nrf9160 nrf9160dk_nrf9160_ns
+    integration_platforms:
+      - nrf9160dk_nrf9160
+      - nrf9160dk_nrf9160_ns
```
[the remark this answers is not in the document]
Comment (chmielewskiandreas) [reply tests/subsys/net/lib/download_client/testcase.yaml:7]: I just copied the yaml file from fota_download, but I can add qemu_cortex_m3

File: tests/subsys/net/lib/download_client/src/main.c (hunk 2 above)
Comment (chmielewskiandreas) [reply tests/subsys/net/lib/download_client/src/main.c:93]: actually it is not needed anymore as I just have tested.

File: tests/subsys/net/lib/download_client/testcase.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (simensrostad) [reply tests/subsys/net/lib/download_client/testcase.yaml:7]: Consider adding support for `native_posix` as well to be able to generate coverage reports with twister.

```
./zephyr/scripts/twister --coverage -p native_posix -T ./nrf/tests/subsys/net/lib/download_client/
```

**Review by simensrostad — COMMENTED** [review]
I wonder, has `Unity + CMock` been considered for this test? CMock autogenerates mocks for APIs which removes the need of having test specific mocking libraries. I think it scales better and could possibly prevent a lot of upkeep in the long run.

**Comment by chmielewskiandreas** [comment]
> I wonder, has `Unity + CMock` been considered for this test? CMock autogenerates mocks for APIs which removes the need of having test specific mocking libraries. I think it scales better and could possibly prevent a lot of upkeep in the long run.

Yes, I did consider CMock.

My first decision was to cut the interface to http/coap from the download client as you see in dl_coap.h (5 functions) and dl_http.h (2 functions). I do not want to test coap/http internals for now and for the download client it actually should not matter what kind of data it is receiving.

Unfortunately functions like coap_parse (dl_coap.c) use and modify structure of the download client. Therefore it is not easily possible to use CMock here. I also did not want to refactor the download client for now to make it testable.

The main reason not to use CMock is that I need to create a stack of e.g. different return values of a mocked function within a single test. Take a look at function main.c->test_download_reconnect_on_socket_error. I start a download of 75 bytes that I receive via 3 times 25 bytes but the second call of recvfrom return -1. . ```mock_return_values("mock_socket_offload_recvfrom"``` generates for me proper values on the stack. So far I know CMock does not provide this functionality.

**Review by rlubos — APPROVED** [review]

**Comment by rlubos** [comment]
Would you mind rebasing? We had some changes on how the CI is executed recently so there is some stall status info on the PR.

File: tests/subsys/net/lib/download_client/prj.conf [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+CONFIG_NET_IPV4=y
+CONFIG_NET_LOG=y
+CONFIG_NET_SOCKETS=y
+CONFIG_NET_SOCKETS_OFFLOAD=y
+CONFIG_MINIMAL_LIBC_MALLOC_ARENA_SIZE=2048
+
+CONFIG_COAP=n
+
+CONFIG_LOG=y
+CONFIG_LOG_MODE_IMMEDIATE=y
```
Comment (lemrey) [inline tests/subsys/net/lib/download_client/prj.conf:23]: There's `TEST_LOGGING_DEFAULTS`, it should be the `default` for ztests already

File: tests/subsys/net/lib/download_client/src/main.c [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+	struct download_client client;
+	int32_t recvfrom_params[] = { 25, -1, 25, 25 };
+	int32_t sendto_params[] = { 20, 20, 20, 20 };
+
+	dl_coap_init(75, 20);
+
+	mock_return_values("mock_socket_offload_recvfrom", recvfrom_params,
+			   ARRAY_SIZE(recvfrom_params));
+	mock_return_values("mock_socket_offload_sendto", sendto_params,
+			   sizeof(recvfrom_params) / sizeof(sendto_params[0]));
```
Comment (lemrey) [inline tests/subsys/net/lib/download_client/src/main.c:140]: Did you mean `ARRAY_SIZE(sendto_params)`?

File: tests/subsys/net/lib/download_client/src/main.c [hunk 7]
```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+#include <logging/log.h>
+LOG_MODULE_REGISTER(download_client_test, LOG_LEVEL_DBG);
```
Comment (lemrey) [inline tests/subsys/net/lib/download_client/src/main.c:7]: Is this a leftover? You don't seem to log in this file

File: tests/subsys/net/lib/download_client/src/main.c [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+	struct download_client client;
+	int32_t recvfrom_params[] = { 25, 25, 25 };
+	int32_t sendto_params[] = { 20, 20, 20 };
+
+	dl_coap_init(75, 20);
+
+	mock_return_values("mock_socket_offload_recvfrom", recvfrom_params,
+			   ARRAY_SIZE(recvfrom_params));
+	mock_return_values("mock_socket_offload_sendto", sendto_params,
+			   sizeof(recvfrom_params) / sizeof(sendto_params[0]));
```
Comment (lemrey) [inline tests/subsys/net/lib/download_client/src/main.c:120]: Did you mean `ARRAY_SIZE(sendto_params`) ?

**Review by lemrey — COMMENTED** [review]
Left a few comments.
Had not seen `ztest` being used like this before, it's pretty cool, thanks for contributing!

**Comment by chmielewskiandreas** [comment]
> Would you mind rebasing? We had some changes on how the CI is executed recently so there is some stall status info on the PR.

done :)

File: tests/subsys/net/lib/download_client/testcase.yaml [hunk 9]
```diff
@@ -0,0 +1,7 @@
+tests:
+  net.lib.download_client:
+    tags: fota
+    platform_allow: nrf9160dk_nrf9160 nrf9160dk_nrf9160_ns qemu_cortex_m3
+    integration_platforms:
+      - nrf9160dk_nrf9160
+      - nrf9160dk_nrf9160_ns
```
Comment (gopiotr) [inline tests/subsys/net/lib/download_client/testcase.yaml:7]: Please change `qemu_cortex_m3` to `native_posix` platform and add it also into `integration_platforms`, like this:
```
platform_allow: native_posix nrf9160dk_nrf9160 nrf9160dk_nrf9160_ns
integration_platforms:
  - native_posix
  - nrf9160dk_nrf9160
  - nrf9160dk_nrf9160_ns
```

**Review by gopiotr — CHANGES_REQUESTED** [review]
First - thank you for your contribution :) 

Second - I tried to build and run those tests on `qemu_cortec_m3` simulator, by this command:
```
../zephyr/scripts/twister -v -v -p qemu_cortex_m3 -T tests/subsys/net/lib/download_client
```
And I observed following error during building:
```
...
[ 92%] Linking C executable zephyr_pre0.elf
/home/redbeard/gnuarmemb/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.bfd: subsys/net/ip/libsubsys__net__ip.a(net_if.c.obj): in function `sys_rand32_get':
/home/redbeard/ncs/twister-out/qemu_cortex_m3/net.lib.download_client/zephyr/include/generated/syscalls/rand32.h:44: undefined reference to `z_impl_sys_rand32_get'
/home/redbeard/gnuarmemb/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.bfd: subsys/net/ip/libsubsys__net__ip.a(net_context.c.obj): in function `sys_rand32_get':
/home/redbeard/ncs/twister-out/qemu_cortex_m3/net.lib.download_client/zephyr/include/generated/syscalls/rand32.h:44: undefined reference to `z_impl_sys_rand32_get'
/home/redbeard/gnuarmemb/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.bfd: drivers/net/libdrivers__net.a(slip.c.obj): in function `sys_rand32_get':
/home/redbeard/ncs/twister-out/qemu_cortex_m3/net.lib.download_client/zephyr/include/generated/syscalls/rand32.h:44: undefined reference to `z_impl_sys_rand32_get'
collect2: error: ld returned 1 exit status
make[2]: *** [zephyr/CMakeFiles/zephyr_pre0.dir/build.make:119: zephyr/zephyr_pre0.elf] Error 1
make[1]: *** [CMakeFiles/Makefile2:3183: zephyr/CMakeFiles/zephyr_pre0.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
```
I guess, that some random functions are not available on `qemu_cortex_m3`. I propose change it for `native_posix` platform and everything should works correctly. Then you should be able to run those test by this command (from `nrf` directory):
```
../zephyr/scripts/twister -v -v -p native_posix -T tests/subsys/net/lib/download_client
```
In printed logs, you could observe performed tests:
```
DEBUG   - Spawning BinaryHandler Thread for native_posix/net.lib.download_client
DEBUG   - OUTPUT: [00:00:00.000,000] <err> eth_posix: Cannot create zeth (-1)
DEBUG   - OUTPUT: *** Booting Zephyr OS build v2.7.99-ncs1-2223-g240c7b9a691f  ***
DEBUG   - OUTPUT: Running test suite lib_fota_download_test
DEBUG   - OUTPUT: ===================================================================
DEBUG   - OUTPUT: START - test_download_simple

...

DEBUG   - OUTPUT: [00:00:05.480,000] <dbg> download_client: download_thread: Read 25 bytes from socket
DEBUG   - OUTPUT:  PASS - test_download_abort_on_invalid_block in 0.920 seconds
DEBUG   - OUTPUT: ===================================================================
DEBUG   - OUTPUT: Test suite lib_fota_download_test succeeded
DEBUG   - OUTPUT: ===================================================================
DEBUG   - OUTPUT: PROJECT EXECUTION SUCCESSFUL
```

Third - please add also `native_posix` platform into `integration_platforms`. In simply words: platforms listed in `integration_platforms` store information for our internal CI which platforms should be used during performing CI tests.

File: tests/subsys/net/lib/download_client/testcase.yaml (hunk 9 above)
Comment (chmielewskiandreas) [reply tests/subsys/net/lib/download_client/testcase.yaml:8]: Thanks, I did the changes as you have suggested

**Review by gopiotr — APPROVED** [review]

**Pull request merged**